### PR TITLE
Support all Debug.Assert(...) overloads in flow analysis

### PR DESCRIPTION
**Pull request by buyaa-n**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/5403

Flow analysis doesn't support `Debug.Assert(bool, string? message)` or any other overload except `Debug.Assert(bool)`:
```cs
void M1()
{
    Debug.Assert(OperatingSystem.IsWindows(), "Only supported on windows");
    WindowsOnly(); // warns even not supposed to

    Debug.Assert(OperatingSystem.IsWindows());
    WindowsOnly(); // not warn
}

[SupportedOSPlatform("windows")]
void WindwosOnly () { }
```
I believe all overloads of Debug.Assert(...) should be supported cc @mavasani

### CODE REVIEW

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 1]
```diff
@@ -990,8 +990,11 @@ private bool IsContractCheckMethod(IMethodSymbol method)
             return false;
         }
 
-        private bool IsDebugAssertMethod(IMethodSymbol method)
-            => Equals(method, DebugAssertMethod);
+        private bool IsDebugAssertMethod(IMethodSymbol method) =>
+            DebugAssertMethod != null &&
```
Comment (buyaa-n) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:994]: `DebugAssertMethod` property initialized with `Debug.Assert(bool)` method if exist

Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:994]: Can you please add a doc comment and also rename the method to IsAnyDebugAssertMethod?

**Review by mavasani — APPROVED** [review]

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs [hunk 2]
```diff
@@ -3515,6 +3515,55 @@ Class Test
         M2()
     End Sub
 
+    <SupportedOSPlatform(""Windows10.1.2.3"")>
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:3518]: That might be too much work, but it'd be nice to add tests for the various rules impacted by having this change.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 3]
```diff
@@ -990,8 +990,11 @@ private bool IsContractCheckMethod(IMethodSymbol method)
             return false;
         }
 
-        private bool IsDebugAssertMethod(IMethodSymbol method)
-            => Equals(method, DebugAssertMethod);
+        private bool IsDebugAssertMethod(IMethodSymbol method) =>
```
Comment (Evangelink) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:993]: Would it be worth to add some option to enable/disable this behavior? I absolutely understand the reason of this change but I think that depending on the context of the application we may want to be really sure of some "state". 
I am not sure if I am really clear but `Debug.Assert` gives some safety in `Debug` but after that in `Release` you no longer have any check which means you could end up in this invalid scenario.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 4]
```diff
@@ -990,8 +990,11 @@ private bool IsContractCheckMethod(IMethodSymbol method)
...
+            DebugAssertMethod != null &&
+            method.ContainingSymbol.Equals(DebugAssertMethod.ContainingSymbol, SymbolEqualityComparer.Default) &&
+            method.Name == DebugAssertMethod.Name &&
+            method.ReturnType == DebugAssertMethod.ReturnType;
```
Comment (Youssef1313) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:997]: SymbolEqualityComparer? I'm surprised `==` actually works.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 1 above)
Comment (buyaa-n) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:994]: Sure

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:3518]: > That might be too much work

That is what i thought 😛, ok, i will add some more

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 3 above)
Comment (buyaa-n) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:993]: Not sure I understood your point, but `Debug.Assert` is for asserting expected behavior in `Debug` mode and to catch a failure in a development phase, it should/could not assert `Release`

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs (hunk 4 above)
Comment (buyaa-n) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:997]: Did not undertsood, for `SymbolEqualityComparer` I used `SymbolEqualityComparer.Equals` not `==`

Comment (Youssef1313) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:997]: @buyaa-n I usee you used SymbolEqualityComparer for comparing `ContainingSymbol`s, but not for `ReturnType`s.

> method.ReturnType == DebugAssertMethod.ReturnType;

**Comment by buyaa-n** [comment]
Perf failure unrelated, [filed an issue](https://github.com/dotnet/roslyn-analyzers/issues/5790), mergin the PR

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 5]
```diff
@@ -990,11 +990,19 @@ private bool IsContractCheckMethod(IMethodSymbol method)
             return false;
         }
 
-        private bool IsDebugAssertMethod(IMethodSymbol method)
-            => Equals(method, DebugAssertMethod);
+        /// <summary>
+        /// Checks if the method is an overload of the <see cref="Debug.Assert(bool)"/> method.
+        /// </summary>
+        /// <param name="method">The IMethodSymbol to test.</param>
+        /// <returns>True if the method is an overlaod of the <see cref="Debug.Assert(bool)"/> method.</returns>
```
Comment (MartyIX) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs:997]: `/// <returns>True if the method is an overload of the <see cref="Debug.Assert(bool)"/> method.</returns>` :)

**Pull request merged**